proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1071 @@
+
+    // The resource's logical version. It is illegal to have the same named xDS
+    // resource name at a given version with different resource payloads.
+    string version = 2;
+
+    // The resource payload, including type URL.
+    google.protobuf.Any resource = 3;
+  }
+
+  oneof resource_specifier {
+    option (validate.required) = true;
+
+    // A xdstp:// URL describing how the member resource is to be located.
+    string locator = 1;
+
+    // The resource is inlined in the list collection.
+    InlineEntry inline_entry = 2;
+  }
+}
+```
+
+A simple example of a listener collection fetch sequence:
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a list of resources: `[locator:
+   xdstp://some-authority/envoy.config.listener.v3.Listener/bar, locator:
+   xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+3. Client requests both
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz.`
+4. Server responds with resources contents for
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz`.
+
+This involves two round-trips. Inlining of resources is supported, which can reduce RTT, i.e.
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a collection of literal resources inlined in `foo`:
+   `[inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/bar,
+   inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+
+Note that in all cases, the context parameters presented in the original request
+must be present in the response. For example, a request for
+`xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo?some=thing`, must
+be responded to with a resource named
+`xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo?some=thing`.
+However, individual resources referenced by locator inside the returned list
+collection do not need to have the collection-level context parameters included
+in their naming.
+
+#### Glob
+
+Having explicit collection resources allows delta xDS to provide delta updates
+on individual resources but the resource list itself can become a bottleneck for
+frequently updated large collections of resources, e.g. with 1M endpoints updating
+every 10s.  For this situation, we introduce *glob
+collections*, which allow the `<T>Collection` directory to be elided when
+resources in a collection share a directory structure and a `/*` component is appended to
+the resource path. Continuing the previous example, this now looks like:
+
+1. Client requests
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/foo/*.`
+2. Server responds with resources
```

Comment:
This is a good question. I think we have a choice between not responding and responding with a sentinel value. I think not responding might delay warming, which will matter for something like LEDS. How about we allow a `Resource` to be returned with the `Any` field not set and the name of the glob collection as a sentinel here?

@markdroth thoughts?

Reply:
IMO explicit is better as it allows distinguishing between timeout and non-existence, and may simplify things later with alternatives (in case of a timeout the alternative will be used, but it probably isn't the desired behavior).

Reply:
I've added a proposal to use sentinel empty resource,